drivers/adc/adc_stm32.c

```diff
@@ -0,0 +1,407 @@
+/*
+ * Copyright (c) 2018 Kokoon Technology Limited
+ * Copyright (c) 2019 Song Qiang <songqiang1304521@gmail.com>
+ * Copyright (c) 2019 Endre Karlson
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+
+#include <adc.h>
+#include <device.h>
+#include <kernel.h>
+#include <init.h>
+
+#define ADC_CONTEXT_USES_KERNEL_TIMER
+#include "adc_context.h"
+
+#define LOG_LEVEL CONFIG_ADC_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(adc_stm32);
+
+#include <clock_control/stm32_clock_control.h>
+
+#include "adc_stm32.h"
+
+/* 16 external channels. */
+#define STM32_CHANNEL_COUNT		16
+
+struct adc_stm32_data {
+	struct adc_context ctx;
+	struct device *dev;
+	u16_t *buffer;
+	u16_t *repeat_buffer;
+
+	u8_t resolution;
+	u8_t channel_count;
+#if defined(CONFIG_SOC_SERIES_STM32F0X) || defined(CONFIG_SOC_SERIES_STM32L0X)
+	s8_t acq_time_index;
+#endif
+};
+
+struct adc_stm32_cfg {
+	ADC_TypeDef *base;
+
+	void (*irq_cfg_func)(void);
+
+	struct stm32_pclken pclken;
+	struct device *p_dev;
+};
+
+static int check_buffer_size(const struct adc_sequence *sequence,
+			     u8_t active_channels)
+{
+	size_t needed_buffer_size;
+
+	needed_buffer_size = active_channels * sizeof(u16_t);
+
+	if (sequence->options) {
+		needed_buffer_size *= (1 + sequence->options->extra_samplings);
+	}
+	if (sequence->buffer_size < needed_buffer_size) {
+		LOG_ERR("Provided buffer is too small (%u/%u)",
+				sequence->buffer_size, needed_buffer_size);
+		return -ENOMEM;
+	}
+	return 0;
+}
+
+static void adc_stm32_start_conversion(struct device *dev)
+{
+	const struct adc_stm32_cfg *config = dev->config->config_info;
+
+	LOG_INF("Starting conversion");
+
+	adc_stm32_start_conversion_soc(config->base);
+}
+
+static int start_read(struct device *dev, const struct adc_sequence *sequence)
+{
+	struct adc_stm32_data *data = dev->driver_data;
+	const struct adc_stm32_cfg *cfg = dev->config->config_info;
+
+	int err;
+
+	switch (sequence->resolution) {
+	case 6:
+		break;
+	case 8:
+		break;
+	case 10:
+		break;
+	case 12:
+		break;
+	default:
+		LOG_ERR("Invalid resolution");
+		return -EINVAL;
+	}
+
+	u32_t channels = sequence->channels;
+
+	data->buffer = sequence->buffer;
+	u8_t index;
+
+	index = find_lsb_set(channels) - 1;
+	adc_stm32_setup_channels_soc(cfg->base, index, data->channel_count);
+	adc_stm32_setup_sequence_length_soc(cfg->base, 1);
+	data->channel_count = 1;
+
+	err = check_buffer_size(sequence, data->channel_count);
+	if (err) {
+		return err;
+	}
+
+	adc_stm32_setup_resolution_soc(cfg->base, sequence->resolution);
+	adc_stm32_enable_eoc_soc(cfg->base);
+
+	adc_context_start_read(&data->ctx, sequence);
+	err = adc_context_wait_for_completion(&data->ctx);
+	adc_context_release(&data->ctx, err);
```

Comment:
Remove this line. The lock is now released in `adc_stm32_read[_sync]`.

Reply:
Right.